// @idealic/poker-engine/src/game/validation.ts
import { Game } from '../Game';
import type { Player } from '../types';
import { getCurrentPlayerIndex } from './position';

/**
 * Determines if a player can act (not folded, not all-in)
 */
export function isPlayerEligibleToAct(player: Player): boolean {
  return !player.hasFolded && !player.isAllIn && !player.isInactive;
}

export function isPlayerActive(player: Player): boolean {
  return !player.hasFolded && !player.isInactive;
}

/**
 * Determines if it's a specific player's turn to act.
 * A player can act if they are the first non-acted, non-folded player in position.
 * @param table Current table state
 * @param playerIndex Index of the player to check
 * @returns True if it's the specified player's turn
 */
export function isPlayerTurn(game: Game, playerIndex: number): boolean {
  const activePlayer = getCurrentPlayerIndex(game);
  return activePlayer === playerIndex;
}

/**
 * Determines if a player can check.
 * A player can check if they've matched the highest bet and haven't acted yet.
 * @param table Current table state
 * @param playerIndex Index of the player to check
 * @returns True if the player can check
 */
export function canCheck(game: Game, position: number): boolean {
  const player = game.players[position];
  if (player.hasFolded) return false;

  // On new streets (bet = 0), ignore hasActed flag if current bet is also 0
  const isNewStreetAction = game.bet === 0 && player.currentBet === 0;
  if (!isNewStreetAction && player.hasActed) return false;

  // Can only check if we've matched the current bet
  return player.currentBet === game.bet;
}

/**
 * Determines if a player can call.
 * A player can call if they face a bet higher than their current bet and have enough chips.
 * @param table Current table state
 * @param playerIndex Index of the player to check
 * @returns True if the player can call
 */
export function canCall(game: Game, position: number): boolean {
  const player = game.players[position];
  if (player.hasFolded) return false;

  // On new streets (bet = 0), ignore hasActed flag if current bet is also 0
  const isNewStreetAction = game.bet === 0 && player.currentBet === 0;
  if (!isNewStreetAction && player.hasActed) return false;

  // Can only call if there are existing bets
  if (game.bet === 0) return false;

  // Can only call if we haven't matched the current bet
  if (player.currentBet >= game.bet) return false;

  // Can't call if we don't have enough chips
  const amountToAdd = game.bet - player.currentBet;
  if (amountToAdd >= player.stack) return false;

  return true;
}

/**
 * Determines if a player can bet.
 * A player can bet if there are no existing bets and they have chips.
 * @param table Current table state
 * @param playerIndex Index of the player to check
 * @returns True if the player can bet
 */
export function canBet(game: Game, position: number): boolean {
  const player = game.players[position];
  if (player.hasFolded) return false;

  // On new streets (bet = 0), ignore hasActed flag if current bet is also 0
  const isNewStreetAction = game.bet === 0 && player.currentBet === 0;
  if (!isNewStreetAction && player.hasActed) return false;

  // Can only bet if there are no existing bets
  if (game.bet > 0) return false;

  // Can't bet if we don't have any chips
  if (player.stack <= 0) return false;

  return true;
}

/**
 * Determines if a player can raise.
 * A player can raise if they face a bet and have enough chips.
 * @param table Current table state
 * @param playerIndex Index of the player to check
 * @returns True if the player can raise
 */
export function canRaise(game: Game, position: number): boolean {
  const player = game.players[position];
  if (player.hasFolded) return false;

  // On new streets (bet = 0), ignore hasActed flag if current bet is also 0
  const isNewStreetAction = game.bet === 0 && player.currentBet === 0;
  if (!isNewStreetAction && player.hasActed) return false;

  // Can only raise if there are existing bets
  if (game.bet === 0) return false;

  // Can't raise if we don't have enough chips for min raise
  const minRaiseAmount = game.bet * 2 - player.currentBet;
  if (minRaiseAmount > player.stack) return false;

  return true;
}

/**
 * Determines if a player can fold.
 * A player can fold if they face a bet higher than their current bet.
 * @param table Current table state
 * @param playerIndex Index of the player to check
 * @returns True if the player can fold
 */
export function canFold(game: Game, position: number): boolean {
  const player = game.players[position];
  if (player.hasFolded) return false;

  // On new streets (bet = 0), ignore hasActed flag if current bet is also 0
  const isNewStreetAction = game.bet === 0 && player.currentBet === 0;
  if (!isNewStreetAction && player.hasActed) return false;

  // Can only fold if there are existing bets to call
  if (game.bet === 0) return false;

  // Can only fold if we haven't matched the current bet
  if (player.currentBet >= game.bet) return false;

  return true;
}
